tooling/miniutil/src/build/terminator.rs

```diff
@@ -146,44 +197,50 @@ impl FunctionBuilder {
         next_val: ValueExpr,
     ) {
         let next_block = self.declare_block();
+        let next_block_kind: BbKind = self.cur_block().kind;
         self.finish_block(compare_exchange(
             dest,
             ptr,
             current,
             next_val,
             bbname_into_u32(next_block),
         ));
-        self.set_cur_block(next_block)
+        self.set_cur_block(next_block, next_block_kind);
     }
 
     pub fn expose_provenance(&mut self, dest: PlaceExpr, ptr: ValueExpr) {
         let next_block = self.declare_block();
+        let next_block_kind = self.cur_block().kind;
         self.finish_block(expose_provenance(dest, ptr, bbname_into_u32(next_block)));
-        self.set_cur_block(next_block)
+        self.set_cur_block(next_block, next_block_kind);
     }
 
     pub fn with_exposed_provenance(&mut self, dest: PlaceExpr, addr: ValueExpr) {
         let next_block = self.declare_block();
+        let next_block_kind = self.cur_block().kind;
         self.finish_block(with_exposed_provenance(dest, addr, bbname_into_u32(next_block)));
-        self.set_cur_block(next_block)
+        self.set_cur_block(next_block, next_block_kind);
     }
 
     pub fn lock_create(&mut self, ret: PlaceExpr) {
         let next_block = self.declare_block();
+        let next_block_kind = self.cur_block().kind;
         self.finish_block(lock_create(ret, bbname_into_u32(next_block)));
-        self.set_cur_block(next_block)
+        self.set_cur_block(next_block, next_block_kind);
```

Comment:
This could be made less repetitive by writing it like
```rust
self.finish_with_next_block(|next_block| lock_acquire(lock_id, bbname_into_u32(next_block)));
```